backends/ebpf/ebpfTable.cpp

```diff
@@ -872,4 +872,169 @@ void EBPFCounterTable::emitTypes(CodeBuilder* builder) {
+    builder->blockStart();
+
+    auto fieldEmitter = [builder](const IR::Type* type, cstring name){
+        auto etype = EBPFTypeFactory::instance->create(type);
+        builder->emitIndent();
+        etype->declare(builder, name, false);
+        builder->endOfStatement(true);
+    };
+
+    if (pvs->elementType->is<IR::Type_Bits>()) {
+        auto type = pvs->elementType->to<IR::Type_Bits>();
+        cstring name = "field0";
+        fieldEmitter(type, name);
+        fieldNames.emplace_back(std::make_pair(name, type));
+    } else if (pvs->elementType->is<IR::Type_Tuple>()) {
+        auto tuple = pvs->elementType->to<IR::Type_Tuple>();
+        int i = 0;
+        for (auto field : tuple->components) {
+            cstring name = Util::printf_format("field%d", i++);
+            fieldEmitter(field, name);
+            fieldNames.emplace_back(std::make_pair(name, field));
+        }
+    } else {
+        BUG("Type for value_set not implemented %1%", pvs->elementType);
+    }
+
+    builder->blockEnd(false);
+    builder->endOfStatement(true);
+}
+
+void EBPFValueSet::emitInstance(CodeBuilder* builder) {
+    builder->target->emitTableDecl(builder, instanceName, TableKind::TableHash,
+                                   keyTypeName, valueTypeName, size);
+}
+
+void EBPFValueSet::emitKeyInitializer(CodeBuilder* builder,
+                                      const IR::SelectExpression* expression,
+                                      cstring varName) {
+    if (fieldNames.size() != expression->select->components.size()) {
+        ::error(ErrorType::ERR_EXPECTED,
+                "Fields number of value_set do not match number of arguments: %1%", expression);
+        return;
+    }
+    keyVarName = varName;
+    builder->emitIndent();
+    builder->appendFormat("%s %s = ", keyTypeName.c_str(), keyVarName.c_str());
+    builder->blockStart();
+
+    // initialize small fields up to 64 bits
+    for (unsigned int i = 0; i < fieldNames.size(); i++) {
+        if (fieldNames.at(i).second->is<IR::Type_Bits>()) {
+            int width = fieldNames.at(i).second->to<IR::Type_Bits>()->width_bits();
+            if (width > 64)
+                continue;
+
+            builder->emitIndent();
+            builder->appendFormat(".%s = ", fieldNames.at(i).first);
+            auto keyExpr = expression->select->components.at(i);
+            if (auto mask = keyExpr->to<IR::Mask>()) {
+                builder->append("((");
+                codeGen->visit(mask->left);
+                builder->append(") & (");
+                codeGen->visit(mask->right);
+                builder->append("))");
+            } else {
+                codeGen->visit(keyExpr);
+            }
+            builder->appendLine(",");
+        }
+    }
+
+    builder->blockEnd(false);
+    builder->endOfStatement(true);
+
+    // init other bigger fields
+    for (unsigned int i = 0; i < fieldNames.size(); i++) {
+        if (fieldNames.at(i).second->is<IR::Type_Bits>()) {
```

Comment:
why do a separate loop? Can't this case be handled in the above loop?

Reply:
I've changed the way that key is initialized, so now it is one loop.